### PR TITLE
PKA 2 authenticated user

### PR DESCRIPTION
**Pull request by shreeramneupane**
Application Controller has a before action `authenticate_user!`, which check up request header `Authorization`. `Authorization` header consist of auth token of Gmail OAuth. It is also required to validate   Gmail domain if environment variable `AUTH_DOMAIN` is define. Finally, if auth token is validate, new user is created with the information response if that user is not previous recorded in database. Then, an instance variable `@current_user` is used to store Authenticated user information.

### CODE REVIEW

File: app/controllers/v1/application_controller.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  # Authenticate user from the auth token provided in the header of request and
+  #   Instance variable @current_user to access the Authenticate user
+  #
+  # @required Authorization [String], Request header containing Gmail Oauth auth token
+  #
+  # @return [User::ActiveRecord_Relation], Active Record Object containing the user details
+  def authenticate_user!
+    token = request.headers['Authorization']
+    auth = Authentication::authenticate_with_oauth(token)
+    raise CustomException::Unauthorized unless auth.code === 200
```
Comment (abhinavmsra) [inline app/controllers/v1/application_controller.rb]: Avoid using `===` in ruby. Intended use of  `===` is explained [here](http://www.tutorialspoint.com/ruby/ruby_operators.htm).

File: lib/authentication.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      end
+    end
+
+    # Verify whether the user details is present in system database and if not present create the new user with avilable details
+    #
+    # @params auth [HTTParty::Response], Hash containing response of Google Oauth validation
+    #
+    # @return [User::ActiveRecord_Relation], Active Record Object containing the user details
+    def find_user(auth)
+      user = User.find_by(uid: auth[:id])
```
Comment (abhinavmsra) [inline lib/authentication.rb]: Code for Database Accesses should not be kept in `lib` folder

**Comment by shreeramneupane** [comment]
:ok:

**Comment by abhinavmsra** [comment]
:+1:

**Comment by sudhirt4** [comment]
:+1:

**Pull request merged**